packages/glimmer-runtime/lib/compiled/opcodes/vm.ts

```diff
@@ -306,6 +313,77 @@ export class EvaluateOpcode extends Opcode {
   }
 }
 
+export class EvaluatePartialOpcode extends Opcode {
+  public type = "evaluate-partial";
+  public compiler: CompileInto & SymbolLookup;
+  public name: CompiledExpression<any>;
+
+  constructor({ name, compiler }: { compiler: CompileInto & SymbolLookup, name: CompiledExpression<any> }) {
+    super();
+    this.name = name;
+    this.compiler = compiler;
+  }
+
+  evaluate(vm: VM) {
+    let reference: PathReference<any> = this.name.evaluate(vm);
+    let name: InternedString = reference.value();
+
+    let { serializedTemplate } = vm.env.lookupPartial([name]);
+    let scanner = new Scanner(serializedTemplate, vm.env);
+    let block = scanner.scanInlineBlock(this.compiler.getSymbolTable());
```

Comment:
It's pretty common for a particular partial callsite to flip-flop between two (or more) partials:

```hbs
{{#each model as |post|}}
  {{partial post.type}}
{{/each}}
```

Here, the partial name might go from "featured", "post", "story", "post", "quote", ... The compilation is expensive, so we shouldn't do it more than once per unique template; we should add a cache here:

```js
export class EvaluatePartialOpcode extends Opcode {
  private cache = new dict<Partial>();

  evaluate(vm: VM) {
    ...
    
    if (!this.cache[name]) {
      // do compilation
    }

    ...
}
```

This is similar to a [polymorphic inline cache](https://en.wikipedia.org/wiki/Inline_caching#Polymorphic_inline_caching).